cloud-minecraft/cloud-bukkit/src/main/java/cloud/commandframework/bukkit/argument/NamespacedKeyArgument.java

```diff
@@ -0,0 +1,350 @@
+                final @NonNull Queue<@NonNull String> inputQueue
+        ) {
+            final String input = inputQueue.peek();
+            final String[] split = input.split(":");
+            final int maxSemi = split.length > 1 ? 1 : 0;
+            if (input.length() - input.replace(":", "").length() > maxSemi) {
+                // Wrong number of ':'
+                return ArgumentParseResult.failure(new NamespacedKeyParseException(
+                        BukkitCaptionKeys.ARGUMENT_PARSE_FAILURE_NAMESPACED_KEY_KEY, input, commandContext
+                ));
+            }
+            try {
+                final NamespacedKey ret;
+                if (split.length == 1) {
+                    if (this.requireExplicitNamespace) {
+                        // Did not provide explicit namespace when option was enabled
+                        return ArgumentParseResult.failure(new NamespacedKeyParseException(
+                                BukkitCaptionKeys.ARGUMENT_PARSE_FAILURE_NAMESPACED_KEY_NEED_NAMESPACE, input, commandContext
+                        ));
+                    }
+                    ret = new NamespacedKey(this.defaultNamespace, split[0]);
+                } else if (split.length == 2) {
+                    ret = new NamespacedKey(split[0], split[1]);
+                } else {
+                    // Too many parts, ie not:valid:input
+                    return ArgumentParseResult.failure(new NamespacedKeyParseException(
+                            BukkitCaptionKeys.ARGUMENT_PARSE_FAILURE_NAMESPACED_KEY_KEY, input, commandContext
+                    ));
+                }
+
+                // Success!
+                inputQueue.remove();
+                return ArgumentParseResult.success(ret);
+            } catch (final IllegalArgumentException ex) {
+                final Caption caption = ex.getMessage().contains("namespace") // stupid but works
+                        ? BukkitCaptionKeys.ARGUMENT_PARSE_FAILURE_NAMESPACED_KEY_NAMESPACE
+                        // this will also get used if someone puts >256 chars
+                        : BukkitCaptionKeys.ARGUMENT_PARSE_FAILURE_NAMESPACED_KEY_KEY;
+                return ArgumentParseResult.failure(new NamespacedKeyParseException(caption, input, commandContext));
+            }
+        }
+
+        @Override
+        public @NonNull List<@NonNull String> suggestions(
+                final @NonNull CommandContext<C> commandContext,
+                final @NonNull String input
+        ) {
+            final List<String> ret = new ArrayList<>();
+            ret.add(this.defaultNamespace + ":");
+            if (!input.contains(":") && !input.isEmpty()) {
+                ret.add(input + ":");
+            }
+            return ret;
+        }
+
+    }
+
+
+    /**
+     * Exception used when {@link Parser} fails.
+     *
+     * @since 1.7.0
+     */
+    public static final class NamespacedKeyParseException extends ParserException {
```

Comment:
This is a note, more so than a request. In `cloud-core` I've started generating `equals`, `hashCode` and `toString` for the various parser exceptions. This means that we can then use stuff like
```java
        assertThat(result.getFailure()).hasValue(new NamespacedKeyArgument.NamespacedKeyParseException(
                caption,
                liberal,
                this.context
        ));
```
`toString()` in this case means that truth can differentiate between identical string-representations even when `equals` fails, which gives a slightly better failure mesage.

However, because we haven't really unit tested `cloud-bukkit`, it's less important here.